U/ssabhlok/telescope patch

sanchitsabhlok opened the pull request:
# Description
Added Instrument bandpass.
Added functionality to separately calculate fluxes and throughputs at instrument entrance, to separate the telescope contribution. 

## Changes Made
1. The `self.instrument_bandpass` and `self.telescope_bandpass` attributes, and other additions pertain to their initialization, and passing these variables where relevant in function headers. 
2. `get_telescope_counts` gives the source counts JUST for the telescope. 
3. `calc_counts` is a helper function that does count calculation for arbitrary bandpasses (separate from full throughput, easy to isolate telescope/instrument bandpasses)
4. `calc_SNR` has an additional option now to calculate the variance breakdown of the SNR calculation 
5. A separate commit adjusts the `test_counts` test, to use the zodi normalization from the toml instead of the function call. This fixes a previous issue with zodi magnitude calcualtions and subsequent testing. The hard coded values being tested have been adjusted for the new test.

## Related Issues
None

## Testing Performed
All tests run and passed. No changes to existing functionality. 

---------------------

**NOTE:** Refer to the [UASAL Development Guide](https://teledocs.space/docs/stp202502_0006) for the PR checklist for guidance as well as other general PR information._

sanchitsabhlok committed:
Added Telescope count calculation

Added functionality to separately calculate fluxes and throughputs at instrument entrance, to separate the telescope contribution. Added Instrument bandpass.

sanchitsabhlok committed:
Changed sky value tests

The calculations inside the test_counts function no longer use an argument for sky magnitudes, but use the toml values instead. The values for the test have also been adjusted.

patrickingraham commented on src/stp_etc_imaging/ExposureTimeSNRCalculator.py:
```diff
@@ -67,6 +67,9 @@ def __init__(self, telescope_name):
         self.sky_counts = 0
         self.well_depth = None
         self.bandpass = SpectralElement(Box1D, amplitude=1, x_0=6000, width=7000)
+        self.telescope_bandpass = SpectralElement(Box1D, amplitude=1, x_0=6000, width=7000)
```
it looks like you're repeating the default values in multiple places. Consider making an if/elif statement for the cases instead and then using variables for each input.

sanchitsabhlok replied:
Can you clarify what the if/elif statement is for? I can make `amplitude=1, x_0=6000, width=7000` into variables and pass them through all three.

sanchitsabhlok replied:
fixed

patrickingraham commented on src/stp_etc_imaging/ExposureTimeSNRCalculator.py:
```diff
@@ -160,12 +150,13 @@ def add_sensor(self, num_curves=1
         
-    # Add mirrors
-    def add_mirror(self, mirror_qe_fits_file, num_curves=1 ,wave_unit='nm', support_data_path=None, plot=False,plot_title="Mirror"):
+    def add_mirror(self, mirror_qe_fits_file, num_curves=1 ,wave_unit='nm', support_data_path=None, plot=False,
```
i should probably grumble at you for not having docstrings for methods. Specifically for the keywords that aren't obvious (e.g. `num_curves`)

sanchitsabhlok replied:
So this has just carried on since the ETC has existed. I can add a docstring for this method but it might stand out coz none of the other methods have docstrings. Maybe this should be its own future issue? Skipping for now but I'll open an issue.

sfrinaldi replied:
https://github.com/uasal/stp_etc_imaging/issues/12 (issue in question)

patrickingraham commented on src/stp_etc_imaging/ExposureTimeSNRCalculator.py:
```diff
@@ -175,15 +166,16 @@ def add_mirror(self, mirror_qe_fits_file, num_curves=1 ,wave_unit='nm', support_
-    # Add filter properties
-    def add_filter(self, filter_fits_file, wave_unit='angstrom', num_curves=1, support_data_path=None, plot=False, plot_title="Filter"):
+    def add_filter(self, filter_fits_file, wave_unit='angstrom', num_curves=1, support_data_path=None, plot=False,
+                   plot_title="Filter"):
```
should the default title be the name of the fits file?

sfrinaldi replied:
I think so as well or at least be included in the name ^ (Just adding to this)

sanchitsabhlok replied:
Usually this is triggered for all plots (mirror/sensor/telescope) throughput curves so the goal is to distinguish each component from another, not necessarily indicate which file the filter is. I would also personally argue in the case of a filter a user can eyeball which band they're looking at.

patrickingraham commented on src/stp_etc_imaging/ExposureTimeSNRCalculator.py:
```diff
@@ -237,18 +230,10 @@ def calc_PSF(self, wavelength=None, approx_type='sq', verbose=False):
-                 , support_data_dir = None
-                 # , left=4000 ,right=8000  # for plotting (units in angstroms)
-                 ):
+    def make_STP(self, plot=False, filters=True, sensor=True, mirrors=True, bg_sb = None, custom_toml_dir = None,
```
did you run black (or ruff format) on this before/after?

Seems like there are changes that aren't just your code changes

sanchitsabhlok replied:
Uhh this was more of a formatting change done manually, I didn't run anything on the code as such.

patrickingraham commented on src/stp_etc_imaging/ExposureTimeSNRCalculator.py:
```diff
@@ -518,76 +503,104 @@ def make_observation(self, flux=0, flux_units=u.ABmag, bg_flux=None, bg_flux_uni
                     print("Error: Could not calculate Background magnitude.")
                     exit()
-            bg_flux = self.bg_magnitude
+            bg_mag = self.bg_magnitude
```
just the line above, why do you exit instead of raise an exception?

sanchitsabhlok replied:
fixed

patrickingraham replied:
I think you want to throw a warning rather than just a print statement
`import warnings

warning.warn ("blah blah blah")

patrickingraham reviewed (changes requested):
Thanks. Made a few comments.
One thing that needs to be added is a new version notes file. The one in the `docs` folder is with respect to the original version, but now we need a `version.rst` file there as well I think.

sfrinaldi commented:
Just did a local test and looks like everything is fine / working though (was double checking). I might want to do some adjustments in the future on the pyproject.toml (ran into a very specific issue with using a new environment for the local test), but that can wait / don't see it actually being an issue.

sfrinaldi commented:
Think there were some other updates pending anyways for this repo (upcoming / future) but checking- Is this PR still in work / just backlogged? (Or is this stale and can be closed out)

sanchitsabhlok commented:
Actively working on this now. Could merge this and push other changes from
a new PR

On Wed, Feb 11, 2026, 2:16 AM sfrinaldi ***@***.***> wrote:

> *sfrinaldi* left a comment (uasal/stp_etc_imaging#11)
> <https://github.com/uasal/stp_etc_imaging/pull/11#issuecomment-3883194841>
>
> Think there were some other updates pending anyways for this repo but
> checking- PR still in work / just backlogged for now I'm guessing correct?
> (Or is this stale)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uasal/stp_etc_imaging/pull/11#issuecomment-3883194841>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AO5XLUUIBHPARAHD67USZP34LLXNFAVCNFSM6AAAAACMEZWDW6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTQOBTGE4TIOBUGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

sfrinaldi commented:
@patrickingraham Thoughts on this? (Sound good for merging this and pushing the other changes per https://github.com/uasal/stp_etc_imaging/pull/11#issuecomment-3884989225?)

sanchitsabhlok commented:
@sfrinaldi I'm cleaning up my local end. Will try to push all requested changes by End of day and open a re-review. Can merge this then.

sanchitsabhlok committed:
Memory leak Housekeeping

Freeing up function scope variables before returning. Plugging memory leak.

sanchitsabhlok committed:
Fixed spectral defaults hard coding

Created new internal variables to track the spectral range over which throughputs are generated for telescope/instrument.

sanchitsabhlok committed:
Fixed BG calculation

Function no longer exits if calculation can't be made but defaults to a hard coded value (as a last safeguard, this should almost never be triggered)

sanchitsabhlok commented:
Re requested review @patrickingraham  

All tests pass. Comments fixed. Good to merge and will open a new PR for jitter and HWK sensor updates.

patrickingraham commented on src/stp_etc_imaging/ExposureTimeSNRCalculator.py:
```diff
@@ -516,78 +503,113 @@ def make_observation(self, flux=0, flux_units=u.ABmag, bg_flux=None, bg_flux_uni
+                try:
+                    self.calculate_bg_normalization_magnitude()
+                except:
+                    print("Error: Could not calculate Background magnitude.")
```
Same comment here about exception handling as above. You should raise an exception and not just exit

sanchitsabhlok replied:
Uhh there's print statements all over the ETC. Do you want me to just fix these or create a separate issue to create warnings/messages everywhere? Also, we don't currently have a warnings dependency which we will add if I update this.

patrickingraham commented on tests/test_ETC_initialization.py:
```diff
@@ -121,38 +121,37 @@ def test_counts(telescope):
     if telescope == "UM":
         assert round(obs.source_counts.value) == 37780689208
-        assert round(obs.sky_counts.value) == 47
+        assert round(obs.sky_counts.value,2) == 0.48
```
I'm a bit confused here.  Your call to make observation changed, so presumably the sky counts are now lower, but does this mean you've changed your testing coverage?

sanchitsabhlok replied:
Uhh yeah I'd say so. I think updating the test suite to be less reliant on hard coded values is probably a discussion I need to have with you and Stephanie. For now, just sticking to the current suite.

sanchitsabhlok replied:
I have created a new issue for this [here](https://github.com/uasal/stp_etc_imaging/issues/15). Would prefer to do this in a separate patch.

patrickingraham reviewed (approved):
Please make sure testing ticket to be finished (and reviewed) before the first week of March.
